### PR TITLE
Adding disaggregated data to the UNODC homicide dataset

### PR DESCRIPTION
**Pull request by spoonerf**
Adding lots of extra variables to this dataset so we have homicide disaggregated by age, sex and weapon.

### CODE REVIEW

File: etl/steps/data/garden/homicde/2023-01-04/unodc.meta.yml [hunk 1]
```diff
@@ -0,0 +1,163 @@
+dataset:
```
Comment (lucasrodes) [inline etl/steps/data/garden/homicde/2023-01-04/unodc.meta.yml:1]: Probably a wrong file committed?

File: etl/steps/data/garden/homicide/2023-01-04/unodc.meta.yml [hunk 2]
```diff
@@ -48,15 +48,220 @@ dataset:
 tables:
   unodc:
     variables:
-      counts:
-        title: Homicides
+      firearms_homicide_count:
+        title: Total homicides by firearm
         short_unit: ''
```
Comment (lucasrodes) [inline etl/steps/data/garden/homicide/2023-01-04/unodc.meta.yml:53]: You can skip `short_unit` and just not declare it. I believe this gets rendered to `''` in Grapher.

File: etl/steps/data/garden/homicide/2023-01-04/unodc.py [hunk 3]
```diff
@@ -76,13 +76,55 @@ def harmonize_countries(df: pd.DataFrame) -> pd.DataFrame:
...
+def create_total_df(df_tot: pd.DataFrame) -> pd.DataFrame:
+    """Create the total homicides dataframe where we will have total homicides/homicide rate
+    disaggregated by age and sex
+    """
+    df_tot = df_tot.drop(columns=["region", "subregion", "indicator", "dimension", "category", "source"])
+    # There are some duplicates when sex is unknown so let's remove those rows
+    df_tot = df_tot[df_tot["sex"] != "Unknown"]
+    # Make it more obvious what total age and total sex means
+    df_tot["age"] = df_tot["age"].replace("Total", "All ages")
+    df_tot["sex"] = df_tot["sex"].replace("Total", "Both sexes")
```
Comment (lucasrodes) [inline etl/steps/data/garden/homicide/2023-01-04/unodc.py:106]: You can try to use `pd.Series.map` with argument `na_action="ignore"` instead, I believe it runs faster. Something like:

```python
df_tot["age"] = df_tot["age"].map({"Total", "All ages"}, na_action="ignore")
df_tot["sex"] = df_tot["sex"].map({"Total", "Both sexes"}, na_action="ignore")
```

File: etl/steps/data/garden/homicide/2023-01-04/unodc.py [hunk 4]
```diff
@@ -76,13 +76,55 @@ def harmonize_countries(df: pd.DataFrame) -> pd.DataFrame:
...
+
+
+def create_mechanism_df(df_mech: pd.DataFrame) -> pd.DataFrame:
+    """Create the homicides by mechanism dataframe where we will have  homicides/homicide rate
+    disaggregated by mechanism (e.g. weapon)
+    """
+    df_mech = df_mech.drop(columns=["region", "subregion", "indicator", "dimension", "source", "sex", "age"])
+    df_mech["category"] = df_mech["category"].replace(
+        ["Firearms or explosives - firearms", "Another weapon - sharp object"], ["Firearms", "Sharp object"]
+    )
```
Comment (lucasrodes) [inline etl/steps/data/garden/homicide/2023-01-04/unodc.py:123]: Again, might prefer to use `pd.Series.map`.

File: etl/steps/data/garden/homicide/2023-01-04/unodc.py [hunk 5]
```diff
@@ -76,13 +76,55 @@ def harmonize_countries(df: pd.DataFrame) -> pd.DataFrame:
...
+    # Make the table wider so we have a column for each mechanism
+    df_mech = df_mech.pivot(index=["country", "year"], values=["value"], columns=["category", "unit_of_measurement"])
+    # Make the columns nice
+    df_mech.columns = df_mech.columns.droplevel(0)
+    df_mech.columns = df_mech.columns.map("_".join)
+    df_mech = df_mech.reset_index()
```
Comment (lucasrodes) [inline etl/steps/data/garden/homicide/2023-01-04/unodc.py:129]: Seems that this logic is written twice (for `create_total_df` and `create_mechanism_df`), perhaps you can have an internal function, sth like:

```python
def _some_name(df, drop_columns, pivot_index, pivot_values, pivot_columns):
        df = df.drop(columns=drop_columns)
	df = df.pivot(
		index=pivot_index,
		values=pivot_values,
		columns=pivot_columns,
	)
	# Make the columns nice
	df.columns = df.columns.droplevel(0)
	df.columns = df.columns.map("_".join)
	df = df.reset_index()
        return df
```

So overall, the function could look like:

```python
def create_mechanism_df(df_mech: pd.DataFrame) -> pd.DataFrame:
    """Create the homicides by mechanism dataframe where we will have  homicides/homicide rate
    disaggregated by mechanism (e.g. weapon)
    """
    df_mech = df_mech.drop(columns=["region", "subregion", "indicator", "dimension", "source", "sex", "age"])
    df_mech["category"] = df_mech["category"].map({
        "Firearms or explosives - firearms": "Another weapon - sharp object",
        "Firearms": "Sharp object"
    }, na_action="ignore")
    df_mech = _create_df(
    	df_mech,
    	["region", "subregion", "indicator", "dimension", "source", "sex", "age"]
        ["country", "year"],
        ["value"],
        ["category", "unit_of_measurement"]
    )
    return df_mech
```

**Review by lucasrodes — APPROVED** [review]
Some minor suggestions, but LGTM!

File: etl/steps/data/garden/homicide/2023-01-04/unodc.py (hunk 5 above)
Comment (spoonerf) [reply etl/steps/data/garden/homicide/2023-01-04/unodc.py:129]: Oo thank you!

File: etl/steps/data/garden/homicide/2023-01-04/unodc.py (hunk 3 above)
Comment (spoonerf) [reply etl/steps/data/garden/homicide/2023-01-04/unodc.py:106]: I think this replaces anything not in the dict with NaN, which I don't want. I just want to replace some of the values in that column with a more descriptive value.

Comment (spoonerf) [reply etl/steps/data/garden/homicide/2023-01-04/unodc.py:106]: Ah if I have ` df_tot["age"].map({"Total", "All ages"}, na_action="ignore").fillna(df_tot['age']` then it fills any non-replaced values with the original column

**Pull request merged**